src/System.Net.Primitives/src/System/Net/CredentialCache.cs

```diff
@@ -222,102 +221,169 @@ public NetworkCredential GetCredential(string host, int port, string authenticat
             {
+                Debug.Assert(cache != null);
                 _cache = cache;
-                _array = new ICredentials[table.Count + hostTable.Count];
-                ((ICollection)table.Values).CopyTo(_array, 0);
-                ((ICollection)hostTable.Values).CopyTo(_array, table.Count);
-                _version = version;
+                _version = cache._version;
             }
 
-            object IEnumerator.Current
+            public object Current
             {
                 get
                 {
-                    if (_index < 0 || _index >= _array.Length)
+                    if (!_enumerating)
                     {
                         throw new InvalidOperationException(SR.InvalidOperation_EnumOpCantHappen);
                     }
                     if (_version != _cache._version)
                     {
                         throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
                     }
-                    return _array[_index];
+                    return _current;
                 }
             }
 
-            bool IEnumerator.MoveNext()
+            public bool MoveNext()
             {
                 if (_version != _cache._version)
                 {
                     throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
                 }
-                if (++_index < _array.Length)
+                return _enumerating = MoveNext(out _current);
+            }
+
+            protected virtual bool MoveNext(out NetworkCredential current)
+            {
+                current = null;
+                return false;
+            }
+
+            public virtual void Reset()
+            {
+                _enumerating = false;
+            }
+
+            private class SingleTableCredentialEnumerator<TKey> : CredentialEnumerator
+            {
+                private Dictionary<TKey, NetworkCredential>.Enumerator _enumerator; // mutable struct field deliberately not readonly.
+
+                public SingleTableCredentialEnumerator(CredentialCache cache, Dictionary<TKey, NetworkCredential> table) : base(cache)
                 {
-                    return true;
+                    Debug.Assert(table != null);
+                    _enumerator = table.GetEnumerator();
+                }
+
+                protected override bool MoveNext(out NetworkCredential current) =>
+                    DictionaryEnumeratorHelper.MoveNext(ref _enumerator, out current);
+
+                public override void Reset()
+                {
+                    DictionaryEnumeratorHelper.Reset(ref _enumerator);
+                    base.Reset();
                 }
-                _index = _array.Length;
-                return false;
             }
 
-            void IEnumerator.Reset()
+            private sealed class DoubleTableCredentialEnumerator : SingleTableCredentialEnumerator<CredentialKey>
```

Comment:
Because there should be no need for any subclasses of this type and this makes it explicit. Also, the JIT may be able to emit more efficient calls if it knows there aren't any subclasses. I've seen code review feedback to add `sealed` where it can be added in many CoreFX PRs.